### PR TITLE
Remove Type.MakeGenericType dependency from source generation

### LINKED ISSUE #63843: Source generated System.Text.Json serialization uses AOT-unfriendly reflection
### Description

The source generated System.Text.Json serialization should not be using reflection (https://devblogs.microsoft.com/dotnet/try-the-new-system-text-json-source-generator/#comment-9820).

More context: https://github.com/dotnet/runtimelab/issues/635#issuecomment-1013824472

### Reproduction Steps

- Clone NativeAOT sample from https://github.com/dotnet/runtimelab/tree/feature/NativeAOT/samples/HelloWorld
- Replace Program.cs with:
```
using System;
using System.Text.Json;
using System.Text.Json.Serialization;

Player player = new() { Id = 1 };
string json = JsonSerializer.Serialize(player, JsonContext.Default.Player);
Console.WriteLine(json);

internal class Player
{
    public int? Id { get; set; }
}

[JsonSerializable(typeof(Player), GenerationMode = JsonSourceGenerationMode.Serialization)]
internal partial class JsonContext : JsonSerializerContext { }
```
- Add `<PropertyGroup><TrimmerSingleWarn>false</TrimmerSingleWarn></PropertyGroup>` to .csproj file
- `dotnet publish -c Release -r win-x64`

### Expected behavior

No warnings about uses of AOT-unfriendly reflection APIs.

### Actual behavior

/_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/GenericMethodHolder.cs(31): AOT analysis warning IL3050: System.Text.Json.Serialization.Metadata.GenericMethodHolder.CreateHolder(Type): Using member 'System.Type.MakeGenericType(Type[])' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The native code for this instantiation might not be available at runtime.

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Removes dependency on `Type.MakeGenericType` from the runtime components of STJ source generation. Should be backported to release/6.0.

Fix #63843

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultValueHolder.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    /// <summary>
+    /// Helper class used for calculating the default value for a given System.Type instance.
+    /// </summary>
+    internal sealed class DefaultValueHolder
+    {
+        private DefaultValueHolder(Type type)
+        {
+            if (type.IsValueType && Nullable.GetUnderlyingType(type) == null)
+            {
+                DefaultValue = Activator.CreateInstance(type);
```
Comment (eerhardt) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultValueHolder.cs:15]: I think we want to call `RuntimeHelpers.GetUninitializedObject` here instead. See https://github.com/dotnet/runtime/pull/47722 for reasoning.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultValueHolder.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+        /// <summary>
+        /// Returns true if <param name="value"/> contains only default values.
+        /// </summary>
+        public bool IsDefaultValue(object value) => DefaultValue is null ? value is null : DefaultValue.Equals(value);
+
+        /// <summary>
+        /// Creates a holder instance representing a type.
+        /// </summary>
+        public static DefaultValueHolder CreateHolder(Type type) => new DefaultValueHolder(type);
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultValueHolder.cs:32]: Any reason to keep this factory method? i.e. the ctor could just be used

**Review by steveharter — APPROVED** [review]

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultValueHolder.cs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultValueHolder.cs:15]: Presumably we'd still need to check if `type` is a struct though right?

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultValueHolder.cs:15]: That method is not available in `netstandard2.0` 😑 I'm guessing it's equivalent to the old `FormatterServices.GetUninitializedObject` but I'd rather not introduce a dependency on `System.Runtime.Serialization`.

**Comment by steveharter** [comment]
Ci issues:
```
System.Text.Json.Serialization.Metadata.DefaultValueHolder.DefaultValueHolder(Type): '#0' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicParameterlessConstructor' in call to 'System.Activator.CreateInstance(Type)'. The parameter 'type' of method 'System.Text.Json.Serialization.Metadata.DefaultValueHolder.DefaultValueHolder(Type)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultValueHolder.cs (hunk 1 above)
Comment (eerhardt) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultValueHolder.cs:15]: > Presumably we'd still need to check if type is a struct though right?

Yes.

> That method is not available in netstandard2.0

You can use the same `#if` as I did in #47722.

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultValueHolder.cs:15]: Cool, I'll do that then, given there is precedent 👍

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultValueHolder.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+namespace System.Text.Json.Serialization.Metadata
+{
+    /// <summary>
+    /// Helper class used for calculating the default value for a given System.Type instance.
+    /// </summary>
+    internal sealed class DefaultValueHolder
+    {
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2067:UnrecognizedReflectionPattern",
+                    Justification = "CreateValueType is only called on a ValueType. You can always create an instance of a ValueType.")]
```
Comment (eerhardt) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultValueHolder.cs:14]: (nit) What is `CreateValueType`?

**Review by eerhardt — APPROVED** [review]
:shipit:

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonParameterInfo.cs [hunk 4]
```diff
@@ -101,14 +101,14 @@ public virtual void Initialize(JsonParameterInfoValues parameterInfo, JsonProper
                 // doesn't match the parameter type, use reflection to get the default value.
                 Type parameterType = parameterInfo.ParameterType;
 
-                GenericMethodHolder holder;
+                DefaultValueHolder holder;
```
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonParameterInfo.cs:104]: Ideally, we shouldn't even have this reflection-related code in code-paths that the source-gen feature flows through. The code for reflection/source-gen should be split/refactored from this [call site](https://github.com/dotnet/runtime/blob/65a5d0e8a3027f2b240d098f47327ba3e1f743b1/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.Cache.cs#L649), but that can be done in a follow up.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs [hunk 5]
```diff
@@ -124,19 +124,12 @@ public partial class JsonTypeInfo
...
         /// </summary>
-        internal GenericMethodHolder GenericMethods
-        {
-            get
-            {
-                _genericMethods ??= GenericMethodHolder.CreateHolder(Type);
-                return _genericMethods;
-            }
-        }
+        internal DefaultValueHolder GenericMethods => _genericMethods ??= DefaultValueHolder.CreateHolder(Type);
```
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs:132]: nit: should the property also be renamed?

**Pull request merged**